server/src/main/java/org/elasticsearch/common/bytes/BytesArray.java

```diff
@@ -67,6 +71,33 @@ public int length() {
         return length;
     }
 
+    @Override
+    public int hashCode() {
+        // NOOP override to satisfy Checkstyle's EqualsHashCode
+        return super.hashCode();
+    }
+
+    @Override
+    protected int calculateHashCode() {
+        int result = 1;
+        for (int i = 0; i < length; i++) {
+            result = 31 * result + bytes[offset + i];
+        }
+        return result;
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        if (this == other) {
+            return true;
+        }
+        if (other instanceof BytesArray) {
```

Comment:
I think this is well worth the optimization to not allocating anything for equals and hashCode since we use `BytesArray` as keys in request caches.